rdt/hyper_transformer.py

```diff
@@ -80,73 +78,62 @@ class HyperTransformer:
-                if a ``dtype`` is not supported by the `HyperTransformer``.
-        """
-        transformers = {}
-        if self.dtypes:
-            dtypes = self.dtypes
-        else:
-            dtypes = [
-                data[column].dropna().infer_objects().dtype
-                for column in data.columns
-            ]
-
-        for name, dtype in zip(data.columns, dtypes):
-            try:
-                kind = np.dtype(dtype).kind
-            except TypeError:
-                # probably category
-                kind = 'O'
-
-            transformer_template = self.dtype_transformers.get(kind)
-            if not transformer_template:
-                raise ValueError('Unsupported dtype: {}'.format(dtype))
-
-            if isinstance(transformer_template, str):
-                transformer_template = self._TRANSFORMER_TEMPLATES[transformer_template]
-
-            if not isinstance(transformer_template, type):
-                transformer = deepcopy(transformer_template)
+    @staticmethod
+    def _field_in_data(field, data):
+        all_columns_in_data = isinstance(field, tuple) and all(col in data for col in field)
+        return field in data or all_columns_in_data
+
+    def _update_field_types(self, data):
+        # get set of provided fields including multi-column fields
+        provided_fields = set()
+        for field in self.field_types.keys():
+            if isinstance(field, tuple):
+                provided_fields.update(field)
             else:
-                transformer = transformer_template()
-
-            transformers[name] = transformer
-
-        return transformers
+                provided_fields.add(field)
+
+        for field in data:
+            if field not in provided_fields:
+                self.field_types[field] = self._DTYPES_TO_DATA_TYPES[data[field].dtype.kind]
+
+    def _fit_field_transformer(self, data, field, transformer):
+        transformer = load_transformer(transformer)
+        transformer.fit(data, field)
+        self._transformers_sequence.append(transformer)
+        output_types = transformer.get_output_types()
+        next_transformers = transformer.get_next_transformers()
+        for (output, output_type) in output_types.items():
+            if output_type in self.transform_output_types:
+                self._output_columns.append(output)
+            else:
+                self._generated_columns.append(output)
+                transformed_data = transformer.transform(data)
+                if output not in self.field_transformers and next_transformers is not None:
+                    self.field_transformers[output] = next_transformers[output]
+                if output in self.field_transformers:
+                    next_transformer = self.field_transformers[output]
+                else:
+                    next_transformer = self._default_transformers[output_type]
+                self._fit_field_transformer(transformed_data, output, next_transformer)
```

Comment:
Maybe we should do this after the `for` loop finishes, so multi-column field transformers can be applied on mutli-column outputs?

* Field `a` is passed to `TransformerA`, which produces output `{'a.b': 'a_data_type', 'a.c': 'numerical'}`
* `field_transformers` specifies `{('a.b', 'c'): TransformerABC}`, so this is applied

This should include the `next_transformers` also, so the following is also possible:

* Field `a` is pased to `TransformerA`, which produces outputs `{'a.b': 'a_data_type', 'a.c': 'another_data_type'}`
* `TransformerA.get_next_transformers()` returns `{('a.b', 'a.c'): TransformerBC}`, so this is applied.

And this actually makes me think, that the following should also be possible (we may want to add this one in a second iteration, though):

* Field `a` is pased to `TransformerA`, which produces outputs `{('a.b', 'a.c'): 'a_data_type'}`
* `data_type_transformers` specifies `{'a_data_type': TransformerABC}`, so this is applied.